unified-runtime/source/adapters/offload/adapter.cpp

```diff
@@ -66,28 +66,40 @@ ur_result_t ur_adapter_handle_t_::init() {
 
 UR_APIEXPORT ur_result_t UR_APICALL urAdapterGet(
     uint32_t, ur_adapter_handle_t *phAdapters, uint32_t *pNumAdapters) {
+  std::mutex InitMutex{};
+
   if (phAdapters) {
-    if (++Adapter.RefCount == 1) {
-      Adapter.init();
+    std::lock_guard Guard{InitMutex};
+
+    // We explicitly only initialize the adapter when outputting it
+    if (!Adapter) {
+      Adapter = new ur_adapter_handle_t_{};
+      Adapter->init();
     }
-    *phAdapters = &Adapter;
+    ++Adapter->RefCount;
```

Comment:
Looks odd, postfix instead?

```suggestion
    Adapter->RefCount++;
```